bgpd/bgp_bfd.c

```diff
@@ -238,6 +238,21 @@ void bgp_peer_bfd_update_source(struct peer *p)
 		changed = true;
 	}
 
+	/* Update interface.
+	 * If singlehop configured, and interface is loopback, do not set it
+	 */
+	if (p->nexthop.ifp && bfd_sess_interface(session) == NULL) {
+		if (bgp_nht_get_oif(p) == p->nexthop.ifp
+		    && bfd_sess_hop_count(session) == BFD_SINGLE_HOP_TTL) {
```

Comment:
I didn't look deep, but at least this check and all similar checks are not correct, because `bfd_sess_hop_count` for single-hop session is 1, not 255.
I will do a deeper review later this week.

Reply:
Hi, actually, I was just looking for a way to distinguish shop from mhop configuration in BGP.
the wording `ttl_configured` was wrong. so I renamed it to `min_ttl_configured·`.

And I made some changes..

I use that thread to ask you a question.
Do you recall about configuring bgp multihop session by using interface attribute and update-source command ?
I wonder if this PR does not help in solving it.

Reply:
@idryzhov 
> I didn't look deep, but at least this check and all similar checks are not correct, because bfd_sess_hop_count for single-hop session is 1, not 255.

This is a bug in the `bfd_sess_set_hop_count` and possibly in `bfd_sess_set_minimum_ttl`. We either need to check for bad values and `assert` or correct them.

I have a patch locally that I didn't have time to test further:

```diff
@@ -660,6 +660,16 @@ void bfd_sess_set_mininum_ttl(struct bfd_session_params *bsp, uint8_t min_ttl)
 
 void bfd_sess_set_hop_count(struct bfd_session_params *bsp, uint8_t min_ttl)
 {
+	/*
+	 * Handle corner cases:
+	 * - 0: (default) single hop.
+	 * - 255: (equivalent to 254) multi hop.
+	 */
+	if (min_ttl == 0)
+		min_ttl = 1;
+	if (min_ttl == 255)
+		min_ttl = 254;
+
 	if (bsp->args.ttl == min_ttl)
 		return;
```

This doesn't seem related with this PR though.

---

The bigger issue is: there are situations where BGP can't tell BFD what is the proper configuration. The three scenarios I can think of are:

1. `iBGP` peer using `update-source loopback` """directly connected""" (should we consider source as loopback as directly connected? theoretically it is at least one hop away, but in practice the OS behaves as multihome and just accepts packet as siglehop - no TTL decrement) . `iBGP` is by default considered multi hop.

   Topology example:

   `[R1] <-> [R2]`

   One way we could hint BFD about the distance is using the BGP neighbor TTL configurations: `neighbor A.B.C.D  ttl-security hops NUMBER` (or `neighbor PEER ebgp-multihop NUMBER` for eBGP)

2. `eBGP` peer using `update-source loopback`. Similar as above: BFD will be configured to use the `loopback` address and single hop (unless `ebgp-multihop` is specified).

3. `iBGP` directly connected (not using loopback). By default BGP will assume `iBGP` is multi hop and will configure a multi hop BFD session even though it is directly connected.

I though about using the BGP TTL configurations for solving the issues above, but the anycast setups might also want to control the `source`. Another idea that came up was to use `profile` to override these values (e.g. single/multi-hop mode, TTL and source). Thoughts?

Reply:
@pguibert6WIND:

Regarding the API, I propose not to use `minimum_ttl` at all and always use `hop_count` functions. The former is confusing with all the value transformations, and the latter is simple and conforms to the BGP multihop configuration which also takes number of hops. I already submitted a PR to remove the former API: https://github.com/FRRouting/frr/pull/9956.

Regarding the problem with update-source, I am thinking, how about we just check the `nexthop.ifp` using `if_is_loopback_or_vrf`? And if the interface is loopback or VRF, we don't set it into the BFD session? It should be easier than introducing `bgp_nht_get_oif`.

@rzalamena:

Items 2 and 3 I think are already covered by `PEER_IS_MULTIHOP`. EBGP is treated as single-hop unless `ebgp-multihop` is configured, IBGP is treated as multi-hop unless it runs on a shared network.

Item 1 is interesting. Some configuration is definitely needed in this case as we can't determine whether the session is single-hop or multi-hop (it depends on the neighbor configuration which we don't know). Maybe we can introduce something like `neighbor PEER ibgp-singlehop` to resemble `neighbor PEER ebgp-multihop`?

Reply:
> @pguibert6WIND:
> 
> Regarding the API, I propose not to use `minimum_ttl` at all and always use `hop_count` functions. The former is confusing with all the value transformations, and the latter is simple and conforms to the BGP multihop configuration which also takes number of hops. I already submitted a PR to remove the former API: #9956.
> 
> Regarding the problem with update-source, I am thinking, how about we just check the `nexthop.ifp` using `if_is_loopback_or_vrf`? And if the interface is loopback or VRF, we don't set it into the BFD session? It should be easier than introducing `bgp_nht_get_oif`.

@idryzhov , I agree to keep one single API only. so go for removing set_mininum_ttl.
For  ``if_is_loopback_or_vrf()`, it is not sufficient. Creating a dummy interface can be done and will not match the function. dummy interfaces are often used to imitate loopback interfaces.

Reply:
Hi.

Regarding the commit about the interface name. I don't think we should do that. When you configure `update-source`, it means that your session is multi-hop. And for multi-hop session the interface is not needed at all, so the change is not necessary.

Regarding the commit adding back the singlehop/multihop command – I am OK with this change. It is especially needed when we use `update-source` and our neighbor is directly connected. In this case, for the neighbor the session is multi-hop and for us the session is single-hop (at least that's how `PEER_IS_MULTIHOP` currently works). So having an ability to manually specify the type of the session is needed.

Reply:
agree. lets keep the single commit.
we just ensure that when single hop is configured, we do not set the interface as outgoing interface.